File: metaTagGenerator.pl
#!/usr/bin/perl -w
#
#Plugin Name: MetaTagGenerator
#Plugin URI:  http://blogs.hiveworks.com/msa/content/binary/wordpress/mtg.tar.gz
#Description: A plugin that generates meta tags out of the nouns of the posts
#Author: Michalis Sarigiannidis
#Version: 1.0
#Author URI: http://blogs.hiveworks.com/msa/
#

use strict;
use Lingua::Wordnet;

my ($WORDS) = @ARGV;

# Remove any non-word character
$WORDS =~ s/\W/ /gi;

# Remove all words with 1 to 3 characters
# This doesn't catch them all, fix it
#$WORDS =~ s/(\W|^|\s)\w{1,3}(\W|$|\s)/ /gi; 

# Convert all characters to lower case
$WORDS =~ tr/[A-Z]/[a-z]/;

# Parse string  to array
my @parsedWords = map { s/\W|\d/ /gi; split; } $WORDS;

# Remove all duplicates
my %seen = ();
my @uniqueWords = sort grep { ! $seen{$_}++ } @parsedWords;

# Reject all words that are not nouns
my $wn = new Lingua::Wordnet;
my @nouns = grep { $wn->lookup_synset($_, "n")>0 && length($_) > 4} @uniqueWords;

# Create comma-separated string
my $nouns = join(", ", @nouns);

print "$nouns";

File: metaTagGenerator.php
<?php
/*
        Plugin Name: MetaTagGenerator
        Plugin URI: http://blogs.hiveworks.com/msa/content/binary/wordpress/mtg.tar.gz
        Description: A plugin that generates meta tags out of the nouns of the posts
        Author: Michalis Sarigiannidis
        Version: 1.0
        Author URI: http://blogs.hiveworks.com/msa/
*/

function filterNouns($text)
{
        $text = preg_replace("/(\W)/", " ", $text);

       $nouns = shell_exec('/usr/bin/perl /htdocs/hiveworks.com/blogs/htdocs/msa/wp-content/plugins/metaTagGenerator.pl "'.$text.'"');
        if (!(isset($nouns) && is_string($nouns)))
		$nouns ='';

        return $nouns;
}

function generateMetaTags($postID)
{
        $postData = get_postdata($postID);
        $text = $postData['Title'].' '.$postData['Excerpt'].' '.$postData['Content'];
        $nouns = filterNouns($text);

	$nouns = 'name="keywords" content="'.$nouns.'"';

        if (!update_post_meta($postID, 'head_meta', $nouns))
                add_post_meta($postID, 'head_meta', $nouns, true);
}

add_action('save_post', 'generateMetaTags');
add_action('edit_post', 'generateMetaTags');
add_action('publish_post', 'generateMetaTags');
?>
